### PR TITLE
Moving the rrgraph_read/ rrgraph_write to libs/rrgraph

### PR DESCRIPTION
**Pull request by Raza-jafari-RS**
#### In this PR the  rrgraph_read/ rrgraph_write are moved to libs/rrgraph
Previously the  rrgraph_read/ rrgraph_write were present in the vpr/src/route which is now moved to  libs/rrgraph/src/base 

#### Motivation and Context
As devised by the mentor

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: libs/librrgraph/CMakeLists.txt
Comment (tangxifan) [inline libs/librrgraph/CMakeLists.txt:18]: Please avoid to create a loop between VPR and this library. This library should not have any dependency on VPR. It should be self-contained.

In order to do this, you need to move all the related header and source files to the ``librrgraph``

Therefore, I think this line should not be modified.

File: vpr/CMakeLists.txt
Comment (tangxifan) [inline vpr/CMakeLists.txt:77]: No need to modify this line. You just need to ensure that ``librrgraph`` is in the list of linked library.

**Review by tangxifan — CHANGES_REQUESTED** [review]
I checked the code changes. I believe that you should move all the relevant data structures required by the rr_graph_reader/writer to the ``librrgraph``.
The standard way is that 
- you go through the header files included by the rr_graph_writer.cpp and rr_graph_reader.cpp. 
- Find out what data structure it requires.
- For each data structure, create a separated header file and move it to ``librrgraph``.

If you feel confused, we can get on a call.

**Comment by vaughnbetz** [comment]
Waiting on a response to @tangxifan before reading this in detail ...

**Comment by tangxifan** [comment]
> Waiting on a response to @tangxifan before reading this in detail ...

Hi @vaughnbetz Sorry for a delay in this PR. 
Actually, @Raza-jafari-RS is now assigned to another technical task related to VTR (developing EDIF/Verilog parser due to a higher priority). 
I have an intern who will join in 3 weeks and take over this work.
If you see an emergency, I can talk to @Raza-jafari-RS and see how to catch your deadline.

**Comment by vaughnbetz** [comment]
That's fine. Thanks @tangxifan

**Comment by tangxifan** [comment]
@vaughnbetz It is my pleasure to introduce @oscarcheng105 who is a summer intern at Rapidsilicon.
During his intern, he will contribute to VTR on the following aspects:

- Finish this PR
- Build unit tests for librrgraph and establish QoR tracking
- Benchmarking the efficiency of RRGraph data structure by comparing to the implementation in [rr_graph_obj.h](https://github.com/lnis-uofu/OpenFPGA/blob/master/vpr/src/device/rr_graph_obj.h)

@vaughnbetz 
If you are o.k. with the plan, would you mind to give him contributor access to the project? Thank you in advance.

**Comment by vaughnbetz** [comment]
Welcome to vtr @oscarcheng105 !  I've added you to the committers list.

**Comment by tangxifan** [comment]
Close this PR as #2101 is addressing this issue.

**Pull request closed without merging**
